Calculation develop v1

Y-Pazniak opened the pull request:
The main work has been finished.
Calculator counts the prices.
Calculator forms the result string.
Calculator is endless.
Calculator can fix users mistake on his feet (user doesn't need to restart the calculator if he made the mistake).

Y-Pazniak committed:
1. CSR structure was created.
2. Singleton bot was created.
3. Base bot functions for checking was added.
4. Bot was registered.

Y-Pazniak committed:
First branch added (choose between eaes and other countries).

Y-Pazniak committed:
Second branch of calculation added (physical or juridical person).

Y-Pazniak committed:
StringBuilder with user's calculation path added
The third step (choose age) added

Y-Pazniak committed:
interface finished, calculator shows variants according on user's choice and remembers users input

Y-Pazniak committed:
Methods names were fixed with verbs

Y-Pazniak committed:
ENUMS instead booleans added

Y-Pazniak committed:
String with options fixed by ENUMs. Options work properly. Need to add result user string (createUsersChoiceString()) and to make null to upper steps (ex-booleanChecker()).
As a matter of fact, I consider a good idea to make another singleton-object for strings creation and take away string-methods from Bot()-class.

Y-Pazniak committed:
Summary user string was fixed. ENUM using instead of if-else.

Y-Pazniak committed:
booleanCleaner fixed - long code removed, shorter was added

Y-Pazniak committed:
The main work has been finished.
Calculator counts the prices.
Calculator forms the result string.
Calculator is endless.
Calculator can fix users mistake on his feet (user doesn't need to restart the calculator if he made the mistake).

MaximPilipeiko commented on src/main/java/repository/strings/Storage.java:
```diff
@@ -0,0 +1,91 @@
+package repository.strings;
+
+public class Storage {
```
`Constants` I think more suitable Name for this class.
But finally I recommend to store "translations" in `resources`. Please, read this article for details https://medium.com/codex/localization-and-resource-bundles-in-java-b746a9f9e577

Y-Pazniak replied:
Renamed.

MaximPilipeiko commented on src/main/java/service/Bot.java:
```diff
@@ -0,0 +1,334 @@
+    //bot's token takes from the environmental variables
+    @Override
+    public String getBotToken() {
+        return System.getenv("custom_by_utilsbor_bot");
```
I think `BOT_TOKEN` is more suitable name for this env

Y-Pazniak replied:
Need to learn how to save env variables during building an artifact. Will remove it next week.

MaximPilipeiko commented on src/main/java/service/Bot.java:
```diff
@@ -0,0 +1,334 @@
+        try {
+            if (update.hasMessage() && inputMessage.hasText()) {
+                sendMessage = checkInputMessage(inputMessage);
+            } else sendMessage = getSorryMessage(inputMessage);
```
Better to use curl brackets even for a single line code.
```suggestion
            } else {
               sendMessage = getSorryMessage(inputMessage);
            }
```

Y-Pazniak replied:
Fixed.

MaximPilipeiko commented on src/main/java/service/Bot.java:
```diff
@@ -0,0 +1,334 @@
+            execute(sendMessage);
+        } catch (
+                TelegramApiException e) {
+            e.printStackTrace();
```
Please, also print smth like this: "Fail to process use request."

Y-Pazniak replied:
Added.

MaximPilipeiko commented on src/main/java/repository/strings/Storage.java:
```diff
@@ -0,0 +1,91 @@
+    public static final String BYN_STRING = " BYN\n";
+    public static final String FAREWELL_STRING = "\nДополнительная информация.\n custom.by - оформление утильсбора;  \n /start - перезапустить калькулятор.";
+
+    public static final String START = "/start";
```
I suggest moving it to a separate class or maybe even enum `Command`

Y-Pazniak replied:
Done.

MaximPilipeiko commented on src/main/java/service/Bot.java:
```diff
@@ -0,0 +1,334 @@
+import repository.steps.*;
+import repository.strings.Storage;
+
+public class Bot extends TelegramLongPollingBot {
```
I think it would be best to split this class into two different classes. The first is Bot (controller), which is responsible for receiving the request, passing data to the service (the second class), and sending the response.
We will discuss it on the call

MaximPilipeiko commented on src/main/java/service/Bot.java:
```diff
@@ -0,0 +1,334 @@
+    }
+
+    private String checkForElectricAutoAge() {
+        if (carAge == null) {
```
What about to use Conditional ternary operator?

Y-Pazniak replied:
Was rewrited.

MaximPilipeiko commented on src/main/java/service/Bot.java:
```diff
@@ -0,0 +1,334 @@
+        return sendMessage;
+    }
+
+    private void cleanBooleans(int stepCleaner) {
```
I think the name of this method does not reflect enough of what this method does

Y-Pazniak replied:
Was renamed to cleanStepsAfterCurrent()

MaximPilipeiko commented on src/main/java/service/Bot.java:
```diff
@@ -0,0 +1,334 @@
+        return createSendMessage(message.getChatId().toString());
+    }
+
+    private String createSummaryString() {
```
For me, 'String' prefix is redundant the return type says enough

Y-Pazniak replied:
Tried to remove all the String prefixes.

MaximPilipeiko commented on src/main/java/service/Bot.java:
```diff
@@ -0,0 +1,334 @@
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append(createUsersChoiceString());
+        switch (countryOrigin) {
+            case EAES -> stringBuilder.append(getStringOptionsForEaes());
```
We have to discuss how to build summary result. I'm not sure that it's optimal now.

Y-Pazniak replied:
I have made it a little better, but I still need to improve this part using reflexion or iteration.

MaximPilipeiko commented on src/main/java/repository/strings/Storage.java:
```diff
@@ -0,0 +1,91 @@
+    public static final String PRICE_PASSENGER_OLDER_7_YEARS = "1225.1";
+
+    //other countries - gasoline engine prices
+    public static final String PRICE_PASSENGER_OTHER_GASOLINE_1000_LESS_3_YEARS = "1652.2";
```
I recommend to move this constant to independent class

Y-Pazniak replied:
Done.

MaximPilipeiko reviewed (changes requested):
Good job!

Y-Pazniak committed:
pom.xml fixed according to 21 Java
packages was extracted from the root
big Storage enum was divided into the small parts
constants was renamed into better (in my opinion) way
into enums constans was divided into groups
main was extracted from the controller package and moved to the root
createSummaryString was renamed

Y-Pazniak committed:
Revert "pom.xml fixed according to 21 Java"

This reverts commit 5b29bcb019128e23acc3917ef646ee1b4d1fb5f6.
